File: lyra/src/main/java/com/fondesa/lyra/annotation/SaveState.java
/*
 * Copyright (c) 2017 Fondesa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.fondesa.lyra.annotation;

import android.os.Bundle;

import com.fondesa.lyra.coder.StateCoder;
import com.fondesa.lyra.coder.utils.StateCoderUtils;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Used to annotate the fields that you want to save/restore.
 * You can pass a custom {@link StateCoder} to save/restore a field, otherwise,
 * a default {@link StateCoder} will be used.
 * <br>
 * The possibilities are:
 * <ul>
 * <li>{@code @StateCoder}: StateCoder.class</li>
 * <li>{@code @StateCoder(CustomCoder.class)}: CustomCoder.class</li>
 * <li>{@code @StateCoder(value = CustomCoder.class)}: CustomCoder.class</li>
 * </ul>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.FIELD)
public @interface SaveState {

    /**
     * Specify the coder to serialize the field into a {@link Bundle} and to
     * deserialize it from a {@link Bundle}.
     * The default coder is {@link StateCoder} without a real implementation.
     * The right default coder will be chosen runtime with {@link StateCoderUtils#getBasicCoderForClass(Class)}.
     *
     * @return concrete instance of {@link StateCoder} to use when a field must be serialized/deserialized
     */
    Class<? extends StateCoder> value() default StateCoder.class;
}

File: lyra/src/main/kotlin/com/fondesa/lyra/annotation/SaveState.kt
package com.fondesa.lyra.annotation

import android.os.Bundle
import com.fondesa.lyra.coder.StateCoder
import com.fondesa.lyra.coder.utils.StateCoderUtils
import kotlin.reflect.KClass

/**
 * Used to annotate the fields that you want to save/restore.
 * You can pass a custom [StateCoder] to save/restore a field, otherwise,
 * a default [StateCoder] will be used.
 * The possibilities are:
 * <ul>
 * <li>`@StateCoder`: StateCoder.class</li>
 * <li>`@StateCoder(CustomCoder.class)`: CustomCoder.class</li>
 * <li>`@StateCoder(value = CustomCoder.class)`: CustomCoder.class</li>
 * </ul>
 *
 * @param value Specifies the coder to serialize the field into a [Bundle] and to
 * deserialize it from a [Bundle].
 * The default coder is [StateCoder] without a real implementation.
 * The right default coder will be chosen runtime with [StateCoderUtils.getBasicCoderForClass].
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FIELD)
annotation class SaveState(val value: KClass<out StateCoder<*>> = StateCoder::class)